lib/puppet/facter_impl.rb

```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+#
+# @api private
+# Default Facter implementation that delegates to Facter API
+#
+
+module Puppet
+  class FacterImpl
+    def initialize
+      require 'facter'
+    end
+
+    def value(fact_name)
+      ::Facter.value(fact_name)
+    end
+
+    def add(name, &block)
+      ::Facter.add(name, &block)
+    end
+
+    def method_missing(method_name, *args, &block)
+      raise NoMethodError, "Facter does not implement :#{method_name}" unless ::Facter.respond_to?(method_name)
+
+      if block_given?
+        ::Facter.send(method_name, *args, &block)
+      else
+        ::Facter.send(method_name, *args)
+      end
```

Comment:
This is slow, might be better to force the definition of each method that we want to expose in Puppet